bulk_indexer_pool_test.go

```diff
@@ -0,0 +1,279 @@
+		bi, _ = pool.Get(context.Background(), key) // Get the same indexer.
+		assert.Equal(t, 1, bi.Items())              // Assert it still has the item.
+		assert.Equal(t, prevBI, bi)                 // Assert its the exact same indexer.
+		pool.Put(key, bi)                           // Return to pool.
+
+		biC := pool.Deregister(key)
+		assert.NotNil(t, biC)
+		assert.Len(t, biC, 1)
+		var count int
+		for range biC {
+			count++
+		}
+		assert.Equal(t, 1, count)
+		// Ensure that the indexer client is always reset.
+		assert.NotContains(t, uniques, "invalid")
+		for id := range parameters {
+			assert.Contains(t, uniques, id)
+		}
+	}
+
+	tc := []struct {
+		guaranteed, localMax, total int
+	}{
+		{1, 2, 2},
+		{1, 2, 10},
+		{1, 10, 10},
+		{1, 10, 100},
+		{2, 10, 100},
+		{10, 20, 200},
+		{100, 200, 500},
+	}
+	for _, testCase := range tc {
+		t.Run("pool "+strings.Join(strings.Fields(fmt.Sprint(testCase)), "_"), func(t *testing.T) {
+			test(t, testCase.guaranteed, testCase.localMax, testCase.total)
+		})
+	}
+}
+
+func TestBulkIndexerPoolCorrectness(t *testing.T) {
+	cfg := BulkIndexerConfig{
+		Client: docappendertest.NewMockElasticsearchClient(t, func(http.ResponseWriter, *http.Request) {}),
+	}
+	require.NoError(t, cfg.Validate())
+	guaranteed := 5
+	localMax := 10
+	total := 20
+	pool := NewBulkIndexerPool(guaranteed, localMax, total, cfg)
+
+	// Test that the local max is respected
+	maxOutKey := "test"
+	pool.Register(maxOutKey)
+	indexers := make(chan *BulkIndexer, localMax)
+	for i := 0; i < localMax; i++ {
+		indexer, err := pool.Get(context.Background(), maxOutKey)
+		require.NoError(t, err)
+		require.NotNil(t, indexer)
+		indexers <- indexer
+	}
+
+	// Ensure that the local max is respected.
+	// This should block until the indexer is returned to the pool.
+	assert.Equal(t, int64(localMax), pool.count(maxOutKey))
+
+	// Ensure other IDs can draw from the pool
+	pool.Register("another_test")
+	for i := 0; i < localMax; i++ {
+		indexer, err := pool.Get(context.Background(), "another_test")
+		require.NoError(t, err)
+		require.NotNil(t, indexer)
+	}
+
+	assert.Equal(t, int64(localMax), pool.count("another_test"))
+
+	done := make(chan struct{})
+	started := make(chan struct{})
+	go func() {
+		defer close(done)
+		var once sync.Once
+		// Try to draw more indexers with another ID. Ensure
+		// that the pool is empty and we block until one is returned.
+		for i := 0; i < localMax; i++ {
+			once.Do(func() { close(started) })
+			indexer, err := pool.Get(context.Background(), maxOutKey)
+			require.NoError(t, err)
+			require.NotNil(t, indexer)
+		}
+	}()
+	select {
+	case <-started:
+		time.Sleep(20 * time.Millisecond) // Wait a bit.
+	case <-time.After(time.Second):
+		t.Fatal("Timed out waiting for goroutine to start")
+	}
+	// Ensure "test" key is maxed out.
+	assert.Equal(t, int64(localMax), pool.count(maxOutKey))
+	assert.Equal(t, pool.leased.Load(), int64(total))
```

Comment:
Since this one is a pre-condition to the test (instead of the actual thing that we are testing), I prefer using `require`.
```suggestion
	require.Equal(t, int64(localMax), pool.count(maxOutKey))
	require.Equal(t, int64(total), pool.leased.Load())
```